### ISSUE TITLE
transition radiation plugin produces NaNs 

### ISSUE THREAD

**Issue by PrometheusPi** [ComputationalRadiationPhysics/picongpu#3090]
Currently the transition radiation plugin introduces with #3003 produces NaNs for example in the LWFA simulations for @OmidZar. This is most likely caused by a division by zero as documented in the [gamma filter section of the read the docs documentation](https://picongpu.readthedocs.io/en/latest/usage/plugins/transitionRadiation.html#gamma-filter).

This (hopefully) can be avoided in the LWFA case by increasing the gamma value of the filter (I will try `5` -> `20`) but this is not a general solution. 

As discussed *offline* with @psychocoderHPC, such divisions by zero should be stopped in the kernel (if case is fine, as long as no extra code is executed and only contribution to summation is stopped).

**Comment by PrometheusPi** [comment]
The issue I have encountered was probably caused by a [`nPhi = 1`](https://github.com/ComputationalRadiationPhysics/picongpu/blob/f495273b3a398fbe45ca08feb72a74d15b7b18be/include/picongpu/param/transitionRadiation.param#L136). Should we catch that or is this wrong-usage?

**Comment by sbastrakov** [comment]
I do not have a good idea about the plugin, but since `nPhi` is a `constexpr` you can control that it is in a valid range at compile-time with `static_assert` (or better a pmacc wrapper around it).

**Comment by finnolec** [comment]
@sbastrakov's idea sounds good.I will try to add an assertion and add a comment to the param files. The same should happen if you just detect in a single azimuth direction `nTheta = 1` due to the way how the observation directions are finally calculated in `transitionradiation.param`.

**Comment by finnolec** [comment]
I think I'll do it without an assert and just program border cases, if `nPhi` or `nTheta` are equal to one, so that it'll automatically use `phiMin` and `thetaMin` as the observation angle.

**Comment by sbastrakov** [comment]
Sure, this looks reasonable.

**Comment by PrometheusPi** [comment]
I found out in some production runs, that the CTR is non-zero even though no electron passed the gamma limit of the gamma filter yet. 
![grafik](https://user-images.githubusercontent.com/5121158/68302211-d7e40100-00a1-11ea-9e10-730cb55e7c4c.png)

Blue line marks onset of CTR. Colormap shows electron energy evolution. Gray lines show CTR output periodicity. And the hight of the gray lines is the gamma limit.

**Comment by PrometheusPi** [comment]
I ran a test based on the default transition radiation example. 
I fixed the filter gamma value to 20 and changed the gamma of the bunch from 30, to 15, to 5. My expectation was, that only the first case would produce non-zero CTR, but all produces radiation. 
@finnolec What did I do wrong?

**Comment by sbastrakov** [comment]
@PrometheusPi is this issue connected to #3105 ?

**Comment by PrometheusPi** [comment]
Most likely - @finnolec could not produce NaNs with the default LWFA setup and fix from #3105 - I will rerun a simulation on taurus/ml to test whether there NaNs occur with this fix.

**Comment by sbastrakov** [comment]
@PrometheusPi thanks. And what about [this issue](https://github.com/ComputationalRadiationPhysics/picongpu/issues/3090#issuecomment-550309271)? Sorry, I later edited my previous message.

**Comment by PrometheusPi** [comment]
@sbastrakov Yes - this issue (= #3090) documented NaN production in the CTR plugin.
This was/will be solved with:
 - #3091 - prevents NaNs due to user fault by allowing single angles
 - #3105 - activate filter (could be linked to NaNs - has to be tested)

The question is, whether there are still more causes for NaNs, since NaNs should be captured in the plugin itself. I have the impression, that the filter will not entirely prevent NaNs - but this feeling is just based on the simulation on taurus, where the filter was not working and the CTR plugin generated NaNs.

**Comment by PrometheusPi** [comment]
So please leave it open for now.

**Comment by finnolec** [comment]
Yes, I couldn't create NaNs with #3105 and the default LWFA setup and a gammafilter which is equal to `filterGamma = 1.0`. This means it should also have considered stationary electrons in the calculations, which were most likely the cause of the NaNs.

**Comment by psychocoderHPC** [comment]
@PrometheusPi Is this issue already solved?

**Comment by PrometheusPi** [comment]
No, it is not solved. But it is quite hard to find a suitable test case. 
Please leave this issue open.

For all users that use this plugin: just apply a gamma filter to the particles. This prevents the NaN creation and usually should be fine when computing transition radiation from high energy electrons.

**Comment by sbastrakov** [comment]
@PrometheusPi perhaps this should be clarified in the `transitionRadiation.param` or elsewhere? Sorry if it already it, just from a short look seemed like not the case

**Comment by PrometheusPi** [comment]
@sbastrakov I agree. (I hoped to find the bug sooner.)

**Comment by psychocoderHPC** [comment]
I moved this PR tp the next release

**Comment by finnolec** [comment]
It's been a while since I used this plugin and I don't really remember this error... But from what I gathered by reading through all the messages here, this error only occurs when you try to calculate the transition radiation without any valid, radiating electrons (valid means that their velocity is larger than ```gamma > 1``` and larger than the filter function). So as long as you have a filter function, which checks for non still standing particles, but includes the accelerated electrons from the simulation (```1 < filterGamma < max velocity from electrons```), this plugin should work as expected.

So the plugin does not catch NaN's by itself to my knowledge. But I think it's usually the user's fault if it even generates NaN's... If the issue is just, that there is no comment like "Don't remove the gamma filter! Still standing electrons generate NaN's in the output of the plugin and transition radiation is scaling by \propto \gamma^2" in the ```.param``` file, it should be fixable :)

**Comment by finnolec** [comment]
> For the transition radiation plugin however, there is no time integration, and (as far as I understand it) the filter is the standard instantaneous filter used e.g. also for other plugins. Is this correct @finnolec? 

Yes there is no time integration. The plugin is called once in every timestep it's used and then it calculates the whole transition radiation spectrum in this timestep. I'm not sure what you mean by "instantaneous filter". I compared the filters from the radiation and the transition radiaton plugin right now and they look very similar to me ([transitionRadiation.param](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/include/picongpu/param/transitionRadiation.param) and [radiation.param](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/include/picongpu/param/radiation.param), respectively [transition radiation ExecuteParticleFilter](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/include/picongpu/plugins/transitionRadiation/ExecuteParticleFilter.hpp) and [radiation ExecuteParticleFilter](https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/include/picongpu/plugins/radiation/ExecuteParticleFilter.hpp)). So the implementation looks quite similar to me in both radiation plugins, but I didn't compare it to other plugins yet.

> If the latter is true, having a gamma > 1.0 filter would prevent considering resting particles. However, I would tend to include this check by default into the plugin, if this is the sole cause for errors. (Since having an `all` filter, would break the plugin.)

I think the idea for the filter in the ```param``` file instead of the plugin directly was, that the user can decide how many slow-moving particles they want to use. This means the user could calculate the transition radiation in a low energy simulation (although I'm not sure how relevant this actually is) and the user can set a large gamma value in the filter for a high energy simulation, to improve calculation time (because low energy electrons don't matter then). But if this flexibility is not needed, it makes sense to put it into the plugin.

**Comment by PrometheusPi** [comment]
Okay - due to the not required time integration, using an approach as the radiation plugin is using is not really needed for the transition radiation. A common filter without explicit particle flag attributes that tracks the result of the filter should be fine.
